REVIEW_COPY.md

```diff
@@ -0,0 +1,312 @@
+    | AppendedIn String
+```
+
+We'll create a JSON encoder for this type so we can tag each beacon element with a data attribute containing its position in the outline. We'll then set up our JavaScript to find all elements with such a data attribute in the DOM and feed their coordinates back to us on each drag event. That will allow us to define a type for drag events containing the positions of our beacons on the screen.
+
+```elm
+type alias DragMsg =
+    { cursor : Coords
+    , beacons : List Beacon
+    }
+
+
+type alias Beacon =
+    ( CandidatePosition, Rect )
+```
+
+Remember the `closestRect` function from the polygon example? It's what we need to find the `CandidatePosition` closest to the cursor! Once we know which candidate position is closest, all we need is a function to that moves a node to its new position in an outline:
+
+```elm
+moveNodeInOutline : String -> CandidatePosition -> List OutlineNode -> List OutlineNode
+moveNodeInOutline = Debug.todo "See full source code for implementation!"
+```
+
+It's a tricky function to write, but it doesn't have much to do with drag & drop and so I'm skipping the implementation here. I include a solution with the [outline editor source code]. If you're interested in some thoughts on how to approach data transformations like these, check out this earlier [post on conversion functions], which includes an example of a similar tree manipulation problem.
+
+## Necessary JavaScript
+
+I promised I'd get back at the JavaScript required to make these examples work. All three examples use the same JavaScript code, because it turns out they have the same needs. In every example there's one or more Html elements on the page that we need to track the position and dimensions of as a drag interaction takes place. What our JavaScript code needs to do is generate events when the mouse gets pressed, moved, and released, and bundle with those events the positions of all elements we want to track. We identify those elements by giving them a data attribute with their 'beacon ID'.
+
+There's tons of ways to write this code and I don't believe mine is particularly insightful, so I'll not reprint it here. The [draggable.js source code] for these examples is available though in case you're interested.
+
+## Conclusion
+
+When we need to perform a complicated task, its natural to start by looking for a library to do the heavy lifting for us. Unfortunately I've not found this approach fruitful in building drag & drop interfaces. I believe this is because the draggables and dropzones that are central to the libraries I've tried were not a great fit for the experiences I tried to build, of which I gave three examples in this post.
+
+I've showed a different approach to building drag & drop UI's. This approach uses 'beacons', elements we mark because we'd like to know their position in the DOM for every drag event. I believe this approach allows us to create better drag & drop experiences with far fewer corner cases.
```

Comment:
So, when I read this my immediate thought was "here is where you should announce the light-weight drag & drop library you're publishing" ... not for nothing :)

Reply:
yeah, I kind of thought that too. Could you go into more details here maybe? Could be like:

1. here's the general pattern: make a beacon
   1. here's what a beacon is
   2. here's what a beacon is not (a draggable)
2. here's where this pattern is valuable
   1. here's why this pattern is better than a draggable
3. here's the takeaway: always/never/sometimes use this pattern

Reply:
Thank you both for your comments! I noodled a bit on this, and think I might have wrongly put the emphasis on the beacons here. Primarily this post is an encouragement to try building drag & drop UIs without a library. Beacons help you do that (as the examples hopefully demonstrate), but that's not the main message. I rewrote the conclusion, taking care to try cover the points you list Brian. Would love your thoughts on this:


> When we need to perform a complicated task it's natural to start by looking for a library to do the heavy lifting for us. For building drag & drop interactions libraries give us draggables and dropzones but they are often a bad fit for drag & drop UIs, of which we've seen three in this post.
> 
> I've showed a different approach to building drag & drop UIs. In this approach we subscribe to drag & drop events telling us what's happening in the DOM, update the state of our model based on those events, then update the screen to reflect the new state of the model. If this sounds familiar it's probably because we're describing nothing more than the Elm architecture. In that sense the take-away from this post is encouragement to roll your own drag & drop. I hope the examples in this post show writing drag & drop logic does not need to be an arduous task, and one preferable over the fiddliness often incurred by draggables and dropzones.
> 
> If you do decide to roll your own drag & drop than I can recommend the use of 'beacon elements'. These are regular DOM elements marked so we can access their location on every drag event. Because beacon elements don't need to _do_ anything any element using any positioning strategy can be a beacon. This passive nature distinguishes beacons from draggables or dropzones, both of which include behavior.
> 
> Are drag & drop libraries always a bad idea? I don't think so. In particular there's libraries for specific drag & drop widgets such as re-arrangable lists. Using those when you can could save you a lot of time. There's probably UIs where draggables and dropzones are precisely the right abstraction. Please send me a note if you ran into one of those, I'd love to learn about it! Drag & drop covers an incredibly broad range of functionality though, and so often an off-the-shelf solution will not be available. In those I'd put serious thought into whether draggables and dropzones are going to help you build your UI or make it harder.

Reply:
Like it overall! I think it'd be a more natural flow to reorder the paragraphs (currently 1, 2, 3, 4) as (1, 4, 3, 2).

Reply:
Oh wow, yeah, that's much better!